File: python/src/dexcost/network_accountant.py
"""NetworkAccountant — a per-task, in-process accumulator of HTTP byte usage.

One instance lives (un-serialised) on each Task. The HTTP adapter calls
``record()`` per call; ``finalize()`` is called once at task end. After
finalize the accountant is frozen — later ``record()`` calls are no-ops, so
late-arriving bytes never mutate already-shipped task aggregates.
"""

from __future__ import annotations

import threading
from typing import Any

# Hosts kept in the per-task `by_host` array after finalize (plus `_other`).
FINALIZE_CAP = 20
# Distinct hosts tracked live during the task before overflow folds into
# `_other` — bounds mid-task memory for pathological many-host workloads.
LIVE_CAP = 500


class NetworkAccountant:
    """Accumulates bytes in/out, call count, and a per-host breakdown.

    v2 adds the *external-bytes-out* split: when ``record(...,
    is_internal=True)`` is called, ``bytes_out`` is still added to the
    raw counters but is NOT added to ``external_bytes_out`` — the basis
    for cloud-egress pricing. ``is_internal=False`` and ``is_internal=None``
    both attribute the bytes as external (the ``None`` case — unresolved
    named host — is the common case and is conservatively billable).
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._bytes_in = 0
        self._bytes_out = 0
        self._external_bytes_out = 0
        self._call_count = 0
        # host -> [calls, bytes_in, bytes_out, external_bytes_out]
        self._hosts: dict[str, list[int]] = {}
        # Overflow bucket once LIVE_CAP distinct hosts are tracked.
        self._other = [0, 0, 0, 0]
        self._frozen = False

    def __copy__(self) -> "NetworkAccountant":
        # threading.Lock is not copyable; a copied task must start a fresh,
        # empty accountant — it must not share or inherit frozen state.
        return NetworkAccountant()

    def __deepcopy__(self, memo: dict) -> "NetworkAccountant":
        return NetworkAccountant()

    def record(
        self,
        host: str,
        bytes_in: int,
        bytes_out: int,
        is_internal: bool | None = None,
    ) -> None:
        """Add one HTTP call's bytes. No-op once finalized.

        ``is_internal`` follows the v1 §4.2 three-valued classification:
        - ``True``  → bytes are intra-VPC / loopback → 0 external bytes.
        - ``False`` → confirmed public IP → all of ``bytes_out`` are external.
        - ``None``  → unresolved named host → treated as external
                      (conservative — over-attribute rather than undercount).
        """
        bytes_in = max(0, bytes_in)
        bytes_out = max(0, bytes_out)
        external_out = 0 if is_internal is True else bytes_out
        with self._lock:
            if self._frozen:
                return
            self._bytes_in += bytes_in
            self._bytes_out += bytes_out
            self._external_bytes_out += external_out
            self._call_count += 1
            key = host or "_unknown"
            entry = self._hosts.get(key)
            if entry is not None:
                entry[0] += 1
                entry[1] += bytes_in
                entry[2] += bytes_out
                entry[3] += external_out
            elif len(self._hosts) < LIVE_CAP:
                self._hosts[key] = [1, bytes_in, bytes_out, external_out]
            else:
                self._other[0] += 1
                self._other[1] += bytes_in
                self._other[2] += bytes_out
                self._other[3] += external_out

    def live_host_count(self) -> int:
        """Number of distinct hosts currently tracked (excludes `_other`)."""
        with self._lock:
            return len(self._hosts)

    def finalize(self) -> dict[str, Any]:
        """Freeze the accountant and return the snapshot for the task fields.

        Returns ``{"bytes_in", "bytes_out", "external_bytes_out",
        "call_count", "by_host"}`` where ``by_host`` is
        ``{"hosts": [...]}`` — the top FINALIZE_CAP hosts by total bytes,
        plus an `_other` bucket summing the rest. Each host entry carries
        ``external_bytes_out`` so the v2 per-host egress cost survives the
        top-N cap.
        """
        with self._lock:
            self._frozen = True
            ranked = sorted(
                self._hosts.items(),
                key=lambda kv: kv[1][1] + kv[1][2],
                reverse=True,
            )
            top = ranked[:FINALIZE_CAP]
            overflow = ranked[FINALIZE_CAP:]

            other = list(self._other)  # [calls, bytes_in, bytes_out, external_out]
            for _host, vals in overflow:
                for i in range(4):
                    other[i] += vals[i]

            # If a real host is literally named "_other" it would collide with
            # the synthetic overflow bucket.  Fold it into `other` so the
            # output list never contains two entries with the same host name.
            top_clean = []
            for item in top:
                if item[0] == "_other":
                    for i in range(4):
                        other[i] += item[1][i]
                else:
                    top_clean.append(item)

            hosts: list[dict[str, Any]] = [
                {
                    "host": host,
                    "calls": vals[0],
                    "bytes_in": vals[1],
                    "bytes_out": vals[2],
                    "external_bytes_out": vals[3],
                }
                for host, vals in top_clean
            ]
            if other[0] > 0:
                hosts.append(
                    {
                        "host": "_other",
                        "calls": other[0],
                        "bytes_in": other[1],
                        "bytes_out": other[2],
                        "external_bytes_out": other[3],
                    }
                )
            return {
                "bytes_in": self._bytes_in,
                "bytes_out": self._bytes_out,
                "external_bytes_out": self._external_bytes_out,
                "call_count": self._call_count,
                "by_host": {"hosts": hosts},
            }

File: go/core/network_accountant.go
// Package core — NetworkAccountant ports python/src/dexcost/network_
// accountant.py to Go. One instance lives per task as an unserialized
// in-process accumulator. The HTTP RoundTripper (in adapters/) calls
// Record per call via the registry below; Finalize is called once at
// task end by the tracker. After finalize the accountant is frozen —
// later Record calls are no-ops, so late-arriving bytes never mutate
// already-shipped task aggregates.
//
// Lives in core rather than adapters because the tracker (also in
// core) needs to reference it on task start/end, and core cannot
// import adapters (that'd be a circular import — adapters imports
// core). Mirrors Python's top-level placement of this file.
package core

import (
	"sort"
	"sync"
)

// FinalizeCap is the number of host entries kept in by_host after finalize
// (plus _other). Mirrors python FINALIZE_CAP.
const FinalizeCap = 20

// LiveCap is the maximum distinct hosts tracked live before overflow folds
// into _other. Bounds mid-task memory for pathological many-host workloads.
const LiveCap = 500

// NetworkAccountant accumulates HTTP byte usage for a single tracked task.
//
// isInternal follows the v1 §4.2 three-valued classification (mirrors the
// netbytes.ClassifyDestination return type):
//
//   - *bool == true  → bytes are intra-VPC / loopback → 0 external bytes.
//   - *bool == false → confirmed public IP → all of bytesOut are external.
//   - nil            → unresolved named host → treated as external
//     (conservative — over-attribute rather than undercount).
type NetworkAccountant struct {
	mu                sync.Mutex
	bytesIn           int64
	bytesOut          int64
	externalBytesOut  int64
	callCount         int64
	hosts             map[string][4]int64 // host → [calls, bytes_in, bytes_out, external_bytes_out]
	other             [4]int64
	frozen            bool
}

// NewNetworkAccountant returns an empty, ready-to-use accountant.
func NewNetworkAccountant() *NetworkAccountant {
	return &NetworkAccountant{hosts: make(map[string][4]int64)}
}

// Record adds one HTTP call's bytes. No-op once finalized.
func (a *NetworkAccountant) Record(host string, bytesIn, bytesOut int64, isInternal *bool) {
	// Clamp negatives — bytes can never be negative.
	if bytesIn < 0 {
		bytesIn = 0
	}
	if bytesOut < 0 {
		bytesOut = 0
	}
	externalOut := bytesOut
	if isInternal != nil && *isInternal {
		externalOut = 0
	}

	a.mu.Lock()
	defer a.mu.Unlock()
	if a.frozen {
		return
	}
	a.bytesIn += bytesIn
	a.bytesOut += bytesOut
	a.externalBytesOut += externalOut
	a.callCount++

	key := host
	if key == "" {
		key = "_unknown"
	}

	if entry, exists := a.hosts[key]; exists {
		entry[0]++
		entry[1] += bytesIn
		entry[2] += bytesOut
		entry[3] += externalOut
		a.hosts[key] = entry
	} else if len(a.hosts) < LiveCap {
		a.hosts[key] = [4]int64{1, bytesIn, bytesOut, externalOut}
	} else {
		a.other[0]++
		a.other[1] += bytesIn
		a.other[2] += bytesOut
		a.other[3] += externalOut
	}
}

// LiveHostCount returns the number of distinct hosts currently tracked
// (excludes the synthetic _other bucket).
func (a *NetworkAccountant) LiveHostCount() int {
	a.mu.Lock()
	defer a.mu.Unlock()
	return len(a.hosts)
}

// NetworkSnapshot is the payload returned by Finalize.
type NetworkSnapshot struct {
	BytesIn          int64
	BytesOut         int64
	ExternalBytesOut int64 // canonical scalar — basis for v2 network_cost_usd
	CallCount        int64
	// ByHost is shaped as {"hosts": [...]} where each entry is
	// {host, calls, bytes_in, bytes_out, external_bytes_out}.
	ByHost map[string]interface{}
}

type hostEntry struct {
	host   string
	values [4]int64
}

// Finalize freezes the accountant and returns the snapshot for the task fields.
//
// ByHost contains the top FinalizeCap hosts by total bytes (bytes_in +
// bytes_out), plus an _other bucket summing the rest. Each host entry
// carries external_bytes_out so v2 per-host egress cost survives the cap.
// If a real host is literally named "_other" it is folded into the
// synthetic overflow bucket — the output never has two entries with the
// same name.
func (a *NetworkAccountant) Finalize() NetworkSnapshot {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.frozen = true

	// Drain hosts into a slice we can sort by (bytes_in + bytes_out) desc.
	ranked := make([]hostEntry, 0, len(a.hosts))
	for host, vals := range a.hosts {
		ranked = append(ranked, hostEntry{host: host, values: vals})
	}
	// Stable insertion-order tie-break isn't guaranteed (map iteration is
	// random), but ties between hosts with identical totals are not a
	// correctness issue — they get the same surfacing in the ranking.
	sort.Slice(ranked, func(i, j int) bool {
		totalI := ranked[i].values[1] + ranked[i].values[2]
		totalJ := ranked[j].values[1] + ranked[j].values[2]
		return totalI > totalJ
	})

	other := a.other
	top := make([]hostEntry, 0, FinalizeCap)
	for idx, item := range ranked {
		if idx < FinalizeCap {
			// Fold a real host literally named "_other" into the synthetic
			// bucket so the output never contains a duplicate.
			if item.host == "_other" {
				for i := 0; i < 4; i++ {
					other[i] += item.values[i]
				}
			} else {
				top = append(top, item)
			}
		} else {
			for i := 0; i < 4; i++ {
				other[i] += item.values[i]
			}
		}
	}

	hosts := make([]map[string]interface{}, 0, len(top)+1)
	for _, e := range top {
		hosts = append(hosts, map[string]interface{}{
			"host":               e.host,
			"calls":              e.values[0],
			"bytes_in":           e.values[1],
			"bytes_out":          e.values[2],
			"external_bytes_out": e.values[3],
		})
	}
	if other[0] > 0 {
		hosts = append(hosts, map[string]interface{}{
			"host":               "_other",
			"calls":              other[0],
			"bytes_in":           other[1],
			"bytes_out":          other[2],
			"external_bytes_out": other[3],
		})
	}

	return NetworkSnapshot{
		BytesIn:          a.bytesIn,
		BytesOut:         a.bytesOut,
		ExternalBytesOut: a.externalBytesOut,
		CallCount:        a.callCount,
		ByHost:           map[string]interface{}{"hosts": hosts},
	}
}

// ---------------------------------------------------------------------------
// Registry — task_id → *NetworkAccountant
// ---------------------------------------------------------------------------
//
// The HTTP RoundTripper sees a task_id (via context resolution) but doesn't
// carry a reference to the Task struct. This registry maps task_id strings
// to accountants so the adapter can record bytes via lookup. The tracker
// is responsible for registering on task start and unregistering on
// finalize (Phase D Task 10). Mirrors rust/src/adapters/network_accountant.rs.

var (
	accountantRegistryMu sync.RWMutex
	accountantRegistry   = map[string]*NetworkAccountant{}
)

// RegisterAccountant registers a task's accountant so the HTTP adapter can
// find it by task_id. Replaces any prior registration for the same id.
func RegisterAccountant(taskID string, accountant *NetworkAccountant) {
	accountantRegistryMu.Lock()
	defer accountantRegistryMu.Unlock()
	accountantRegistry[taskID] = accountant
}

// GetAccountant resolves a task's accountant by task_id. Returns nil when
// no task with that id has been registered.
func GetAccountant(taskID string) *NetworkAccountant {
	accountantRegistryMu.RLock()
	defer accountantRegistryMu.RUnlock()
	return accountantRegistry[taskID]
}

// UnregisterAccountant removes and returns a task's accountant. Called by
// the tracker at task end after Finalize() has snapshotted the bytes onto
// the task. Idempotent — returns nil if already removed.
func UnregisterAccountant(taskID string) *NetworkAccountant {
	accountantRegistryMu.Lock()
	defer accountantRegistryMu.Unlock()
	a := accountantRegistry[taskID]
	delete(accountantRegistry, taskID)
	return a
}

// ResetAccountantRegistryForTests clears the entire registry. Exported
// (with the explicit `ForTests` suffix) so cross-package tests in the
// adapters and integration suites can isolate themselves.
func ResetAccountantRegistryForTests() {
	accountantRegistryMu.Lock()
	defer accountantRegistryMu.Unlock()
	accountantRegistry = map[string]*NetworkAccountant{}
}

